### PR TITLE
Adds MapAssert.containsExactly assertion

### PR DESCRIPTION
**Pull request by mchmielarz**
#56 containsExactly part

### CODE REVIEW

File: src/main/java/org/assertj/vavr/api/AbstractMapAssert.java [hunk 1]
```diff
@@ -192,6 +190,25 @@ public SELF containsOnly(Iterable<Tuple2<KEY, VALUE>> entries) {
...
+     *
+     * @param entries the given entries.
+     * @return {@code this} assertions object
+     * @throws NullPointerException     if the given entries array is {@code null}.
+     * @throws AssertionError           if the actual map is {@code null}.
+     * @throws IllegalArgumentException if the given entries array is empty.
+     * @throws AssertionError           if the actual map does not contain the given entries with same order, i.e. the actual map
+     *                                  contains some or none of the given entries, or the actual map contains more entries than the given ones
+     *                                  or entries are the same but the order is not.
+     */
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/api/AbstractMapAssert.java]: The goal of this assertion is to check entries (in order) only - without checking values? Am I right?

File: src/main/java/org/assertj/vavr/internal/Maps.java [hunk 2]
```diff
@@ -124,6 +127,51 @@ public static Maps instance() {
...
+
+        final Map<K, V> expectedEntries = LinkedHashMap.ofEntries(entries);
+        final Map<K, V> notExpected = actual.filter(entry -> !expectedEntries.contains(entry));
+        final Map<K, V> notFound = expectedEntries.filter(entry -> !actual.contains(entry));
+
+        if (notExpected.isEmpty() && notFound.isEmpty()) {
+            // check entries order
+            int index = 0;
+            for (K keyFromActual : actual.keySet()) {
+                if (!areEqual(keyFromActual, entries[index]._1)) {
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/internal/Maps.java]: Maybe add `areNotEqual` method? It makes reading easier.

File: src/main/java/org/assertj/vavr/internal/Maps.java [hunk 3]
```diff
@@ -153,6 +201,28 @@ public static Maps instance() {
...
     }
 
+    /**
+     * Asserts that the number of entries in the given {@code Map} has the same size as the other array.
+     *
+     * @param info  contains information about the assertion.
+     * @param map   the given {@code Map}.
+     * @param other the group to compare
+     * @throws AssertionError if the given {@code Map} is {@code null}.
+     * @throws AssertionError if the given array is {@code null}.
```
Comment (Opalo) [inline src/main/java/org/assertj/vavr/internal/Maps.java]: Shouldn't this is one and the previous one be an `IllegalArgumentException`?

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Test
+    void should_fail_if_one_of_entries_is_null() {
+        final Map<String, String> actual = LinkedHashMap.of("key1", "value1", "key3", "value3");
+
+        assertThatThrownBy(
+                () -> assertThat(actual).containsExactly(ENTRY1, null)
+        )
+                .isInstanceOf(NullPointerException.class);
```
Comment (Opalo) [inline src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: What about `message` verification here? NPE are sometimes quite hard to debug.

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                        "  <List((key1, value1), (key2, value2))>\n" +
+                        "but some elements were not found:\n" +
+                        "  <LinkedHashMap((key2, value2))>\n" +
+                        "and others were not expected:\n" +
+                        "  <LinkedHashMap((key3, value3))>\n"
+                );
+    }
+
+    @Test
+    void should_fail_if_Map_does_not_contain_all_entries_in_smae_order() {
```
Comment (Opalo) [inline src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: `_same_`

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        final Map<String, String> actual = LinkedHashMap.of("key1", "value1", "key3", "value3");
+
+        assertThatThrownBy(
+                () -> assertThat(actual).containsExactly(ENTRY1, ENTRY2)
+        )
+                .isInstanceOf(AssertionError.class)
+                .hasMessage(
+                        "\nExpecting:\n" +
+                        "  <LinkedHashMap((key1, value1), (key3, value3))>\n" +
+                        "to contain exactly (and in same order):\n" +
```
Comment (Opalo) [inline src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: and in *the* same

File: src/main/java/org/assertj/vavr/internal/Maps.java (hunk 2 above)
Comment (mchmielarz) [reply src/main/java/org/assertj/vavr/internal/Maps.java]: Good point

File: src/main/java/org/assertj/vavr/internal/Maps.java (hunk 3 above)
Comment (mchmielarz) [reply src/main/java/org/assertj/vavr/internal/Maps.java]: I just followed assertj conventions and rather would like to stick with them. However, this is a valid point - it's about verifying an argument of the method.

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java (hunk 5 above)
Comment (mchmielarz) [reply src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: 👍

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java (hunk 4 above)
Comment (mchmielarz) [reply src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: True, but this one comes from Vavr directly when converting entries into `LinkedHashMap` and if I'd like to add a message to the exception, I'd have to do something like
```    private <K, V> LinkedHashMap<K, V> asLinkedMap(Tuple2<? extends K, ? extends V>[] entries) {
        return Try.of(() -> LinkedHashMap.ofEntries(entries))
                .mapFailure(
                    Case($(instanceOf(NullPointerException.class)), t -> {
                        if (t.getMessage() == null || t.getMessage().equals("")) {
                            final NullPointerException npe = new NullPointerException("One of entries is null");
                            npe.setStackTrace(t.getStackTrace()); // or npe.initCause(t);
                            return npe;
                        } else {
                            return t;
                        }
                    }))
                .get();
    }```

I don't know is the gain worth of the effort to maintain the code.

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java (hunk 6 above)
Comment (mchmielarz) [reply src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: To change this, we would have to update assertj-core messages of failed assertions. We can make PR to assertj

File: src/main/java/org/assertj/vavr/api/AbstractMapAssert.java (hunk 1 above)
Comment (mchmielarz) [reply src/main/java/org/assertj/vavr/api/AbstractMapAssert.java]: But to check whether an entry is in the order of the actual map we have to check values as well, right? Checking keys only is not sufficient, I guess.

File: src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java (hunk 4 above)
Comment (Opalo) [reply src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: I see. You can also try something like:
```
if(entries.length != Array.of(entries).filter(Objects::nonNull).length()) {
   // throw meaningful exception here
}
```

Comment (mchmielarz) [reply src/test/java/org/assertj/vavr/api/MapAssert_containsExactly_Test.java]: Yup, that's a nice alternative to check this. Thanks!

**Pull request merged**